overlord/configstate/configcore/homedirs.go

```diff
@@ -0,0 +1,185 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/sandbox/apparmor"
+	"github.com/snapcore/snapd/sysconfig"
+)
+
+// For mocking in tests
+var (
+	osutilEnsureFileState = osutil.EnsureFileState
+	osutilDirExists       = osutil.DirExists
```

Comment:
we try to avoid mocking this kind of functions, we just use dirs.SetRootDir and setup appropriate state. See other code in this package using osutil.Ensure*

Reply:
Mmm... As long as we are talking about unit tests, I think it's appropriate to mock all those functions which do not belong to the code being tested: this helps exercising all code paths in isolation, without bringing in more testing setup which could accidentally alter the test outcome. I would keep this mocking and the existing tests as they are now.

This does not mean that we cannot add some slightly more high level test where such mocking does not occur, if you see the need for them. It makes sense to add one for the happy case, but then we already have the spread tests for that (it checks that the system-params file and the AppArmor tunable file are created with the right contents), so I'm not sure.

If I didn't convince you, please let me know if you really want me to rewrite the current tests, or if it's fine if I add some new one (and in that case, whether it's enough to test the positive case or if you'd want to see all the error paths too).

Reply:
OK, I added a couple of commits to reduce mocking. I kept it where it was hard to replace (for example, getting EnsureFileState() to return an error when the FileState is a MemoryFileState is probably impossible), but I still disagree with the idea that reducing mocking is a good goal. That unit tests heavily depend on the implementation is part of the game, due to the fact that in unit tests one wants to test only the smallest testable unit exactly in order to rule out bugs/flakiness in other code that is not under test right now.

It's also true that typically one makes exceptions to this rule when it comes to filesystem primitives, both because they are harder to mock (not in Golang, though) and because their correctness and reliability are usually taken as granted. But when it comes to our own classes in `osutil`, I'm not sure we should make the same assumptions.

The fact that the unit test code heavily depends on the implementation is good, and it means that if you change the implementation, you should have a second look at tests, too. Mocking here ensures that you know exactly what test files need to be updated: if you modified the implementation of `file.go`, you have to change `file_test.go`, and nothing else (well, possibly besides `export_test.go`). If you don't use extensive mocking, then you might have to modify a multitude of other test files, sometimes in ways that are not obvious to someone who is not familiar with all those tests (and I've hit this situation before with snapd).

I know that there's a good reason behind that (us not having a specific directory for component tests, which would make the decision mock/nomock almost clear-cut), but I think that if the code is well tested in the spread, we should be encourage more mocking in our unit tests.

Reply:
> That unit tests heavily depend on the implementation is part of the game, due to the fact that in unit tests one wants to test only the smallest testable unit exactly in order to rule out bugs/flakiness in other code that is not under test right now.

I don't see the "in order" here as a primary requirement, it depends heavily on the nature and characteristics of the dependency.

 > The fact that the unit test code heavily depends on the implementation is good, and it means that if you change the implementation, you should have a second look at tests, too. 

I think unit tests should also help when doing small-scale refactorings and be reusable for as large as possible class of refactorings, if any refactorting forces to rewrite large part of tests, that's a problem. From that POV when applicable they should check what something does vs how it does it. I understand that the what/how separation is subtle, and there are situations where the detail of the how is a primary concern and it is fair for the tests to dig there.

That's why when the setup is not too onerous or contrived I try to consider whether is possible to avoid mocking, at least for direct dependencies, OTOH as I mentioned there are cases where the direct dependence state is complex and the choreography is kind of forced by the API of the dependence (see daemon vs snapstate for example) and mocking seems very appropriate.